### PR TITLE
Internal commands in `/search/transactions`

### PR DESCRIPTION
**Pull request by piotr-iohk**
Implementation of #50.

### CODE REVIEW

File: src/api/search_transactions.rs [hunk 1]
```diff
@@ -2,75 +2,83 @@ use coinbase_mesh::models::{
...
     &self,
     req: SearchTransactionsRequest,
   ) -> Result<SearchTransactionsResponse, MinaMeshError> {
-    let user_commands = self.fetch_user_commands(&req).await?;
-    let user_commands_len = user_commands.len();
-    let next_offset = req.offset.unwrap_or(0) + user_commands_len as i64;
+    let original_offset = req.offset.unwrap_or(0);
+    let mut offset = original_offset;
+    let mut limit = req.limit.unwrap_or(100);
+    let mut transactions = Vec::new() as Vec<BlockTransaction>;
```
Comment (harrysolovay) [inline src/api/search_transactions.rs:22]: The `as Vec<BlockTransaction>` shouldn't be necessary / should be inferred. If it is not for some reason, then perhaps `Vec::new::<BlockTransaction>()` instead.

File: src/api/search_transactions.rs [hunk 2]
```diff
@@ -2,75 +2,83 @@ use coinbase_mesh::models::{
...
+    // Internal Commands
+    if limit > total_count {
+      // if we are below the limit, fetch internal commands
+      (offset, limit) = adjust_limit_and_offset(limit, offset, txs_len);
+      let internal_commands = self.fetch_internal_commands(&req, offset, limit).await?;
+      let internal_commands_len = internal_commands.len() as i64;
+      let internal_commands_total_count = internal_commands.first().and_then(|ic| ic.total_count).unwrap_or(0);
+      transactions.extend(internal_commands.into_iter().map(|uc| uc.into()));
+      txs_len += internal_commands_len;
+      total_count += internal_commands_total_count;
```
Comment (harrysolovay) [inline src/api/search_transactions.rs:43]: Could/should this be extracted into its own fn and reused above (and possibly below as well)?

File: src/api/search_transactions.rs [hunk 3]
```diff
@@ -2,75 +2,83 @@ use coinbase_mesh::models::{
...
+    }
 
+    let next_offset = original_offset + txs_len;
     let response = SearchTransactionsResponse {
-      transactions: user_commands_bt,
-      total_count: user_commands_total_count,
+      transactions,
+      total_count,
       next_offset: match next_offset {
```
Comment (harrysolovay) [inline src/api/search_transactions.rs:55]: `next_offset.map` instead of the `match`?

File: src/api/search_transactions.rs [hunk 4]
```diff
@@ -80,12 +88,31 @@ impl MinaMesh {
...
+
+    let internal_commands = sqlx::query_file_as!(
+      InternalCommand,
+      "sql/indexer_internal_commands.sql",
+      query_params.max_block,
+      query_params.transaction_hash,
+      query_params.account_identifier,
+      query_params.token_id,
+      query_params.status as Option<TransactionStatus>,
+      query_params.success_status as Option<TransactionStatus>,
```
Comment (harrysolovay) [inline src/api/search_transactions.rs:107]: Why are these `as`es necessary?

File: src/api/search_transactions.rs [hunk 5]
```diff
@@ -97,6 +124,129 @@ impl MinaMesh {
...
+    // Derive transaction_identifier by combining command_type, sequence numbers,
+    // and the hash
+    let transaction_identifier = format!(
+      "{}:{}:{}:{}",
+      internal_command.command_type.to_string().to_case(Case::Snake),
+      internal_command.sequence_no,
+      internal_command.secondary_sequence_no,
+      internal_command.hash
+    );
+    let fee = internal_command.fee.unwrap_or_else(|| "0".to_string());
```
Comment (harrysolovay) [inline src/api/search_transactions.rs:157]:
```suggestion
    let fee = internal_command.fee.unwrap_or("0".to_string());
```

File: tests/search_transactions.rs [hunk 6]
```diff
@@ -25,3 +25,45 @@ async fn search_transactions_specified() -> Result<()> {
...
+}
+
+#[tokio::test]
+async fn search_transactions_internal_command() -> Result<()> {
+  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
+
+  let request = SearchTransactionsRequest {
+    network_identifier: Box::new(NetworkIdentifier::new("mina".to_string(), "mainnet".to_string())),
+    max_block: Some(44),
+    // cspell:disable
```
Comment (harrysolovay) [inline tests/search_transactions.rs:55]:
```suggestion
```

File: tests/search_transactions.rs [hunk 7]
```diff
@@ -25,3 +25,45 @@ async fn search_transactions_specified() -> Result<()> {
...
+    transaction_identifier: Some(Box::new(TransactionIdentifier::new(
+      "CkpZZWqdA87JmPxHA5NmFEQ3qh7pUmqXi9GBWzf4pADtPEHQAeH7M".to_string(),
+    ))),
+    // cspell:enable
```
Comment (harrysolovay) [inline tests/search_transactions.rs:59]:
```suggestion
```

File: tests/search_transactions.rs [hunk 8]
```diff
@@ -25,3 +25,45 @@ async fn search_transactions_specified() -> Result<()> {
...
+
+#[tokio::test]
+async fn search_transactions_internal_command() -> Result<()> {
+  let mina_mesh = MinaMeshConfig::from_env().to_mina_mesh().await?;
+
+  let request = SearchTransactionsRequest {
+    network_identifier: Box::new(NetworkIdentifier::new("mina".to_string(), "mainnet".to_string())),
+    max_block: Some(44),
+    // cspell:disable
+    transaction_identifier: Some(Box::new(TransactionIdentifier::new(
```
Comment (harrysolovay) [inline tests/search_transactions.rs:56]:
```suggestion
    transaction_identifier: Some(Box::new(TransactionIdentifier::new(
    // cspell:disable-next-line
```

**Review by harrysolovay — CHANGES_REQUESTED** [review]
Great to see the pagination! Had a few small comments.

File: src/api/search_transactions.rs (hunk 4 above)
Comment (piotr-iohk) [reply src/api/search_transactions.rs:107]: I'm not sure, but compiler complains without them:
```
unsupported type transaction_status for param #5
```

File: src/api/search_transactions.rs (hunk 3 above)
Comment (piotr-iohk) [reply src/api/search_transactions.rs:55]: I'm not sure about `map` :thinking: but I suppose simple `if` would be enough here...

File: src/api/search_transactions.rs [hunk 9]
```diff
@@ -97,6 +121,129 @@ impl MinaMesh {
...
+      transaction_identifier: Box::new(TransactionIdentifier::new(transaction_identifier)),
+      operations,
+      related_transactions: None,
+      metadata: None,
+    };
+
+    BlockTransaction::new(block_identifier, transaction)
+  }
+}
+
```
Comment (joaosreis) [inline src/api/search_transactions.rs:246]: This logic is common to the block endpoint. Should we extract this and reuse in the block endpoint? (the same applies for the other command types)

Comment (piotr-iohk) [reply src/api/search_transactions.rs:246]: That is definitely a good idea! It seems as a bit larger task though, so in order not to bloat this PR I've created a separate task to address it -> https://github.com/MinaFoundation/MinaMesh/issues/53 :+1:

**Review by joaosreis — APPROVED** [review]

**Review by harrysolovay — APPROVED** [review]

**Pull request merged**